bib/cmd/bootc-image-builder/image.go

```diff
@@ -77,21 +77,113 @@ func Manifest(c *ManifestConfig) (*manifest.Manifest, error) {
+
+		// /var is not allowed, but we need to allow any subdirectories that
+		// are not denied below, so we allow it initially and then check it
+		// separately (in checkMountpoints())
+		"/var": {Deny: false},
+
+		// /var subdir denials
+		"/var/home":     {Deny: true},
+		"/var/lock":     {Deny: true}, // symlink to ../run/lock which is on tmpfs
+		"/var/mail":     {Deny: true}, // symlink to spool/mail
+		"/var/mnt":      {Deny: true},
+		"/var/roothome": {Deny: true},
+		"/var/run":      {Deny: true}, // symlink to ../run which is on tmpfs
+		"/var/srv":      {Deny: true},
+		"/var/usrlocal": {Deny: true},
+	})
+
+	mountpointMinimalPolicy = pathpolicy.NewPathPolicies(map[string]pathpolicy.PathPolicy{
+		// allow all existing mountpoints to support size customizations
+		"/":     {Deny: false, Exact: true},
+		"/boot": {Deny: false, Exact: true},
+	})
+)
+
+func checkMountpoints(filesystems []blueprint.FilesystemCustomization, policy *pathpolicy.PathTrie) error {
+	errs := []error{}
+	for _, fs := range filesystems {
+		if err := policy.Check(fs.Mountpoint); err != nil {
+			errs = append(errs, err)
+		}
+		if fs.Mountpoint == "/var" {
+			// this error message is consistent with the errors returned by policy.Check()
+			// TODO: remove trailing space inside the quoted path when the function is fixed in osbuild/images.
+			errs = append(errs, fmt.Errorf("path '/var ' is not allowed"))
+		}
+	}
+	if len(errs) > 0 {
+		return fmt.Errorf("The following errors occurred while validating custom mountpoints:\n%w", errors.Join(errs...))
+	}
+	return nil
+}
+
+func checkFilesystemCustomizations(fsCustomizations []blueprint.FilesystemCustomization, ptmode disk.PartitioningMode) error {
+	var policy *pathpolicy.PathTrie
+	switch ptmode {
+	case disk.BtrfsPartitioningMode:
+		// btrfs subvolumes are not supported at build time yet, so we only
+		// allow / and /boot to be customized when building a btrfs disk (the
+		// minimal policy)
+		policy = mountpointMinimalPolicy
+	default:
+		policy = mountpointPolicy
+	}
+	if err := checkMountpoints(fsCustomizations, policy); err != nil {
 		return err
 	}
-	if customFS := customizations.GetFilesystems(); len(customFS) != 0 {
-		allowedMounts := []string{"/", "/boot"}
-		for _, mp := range customFS {
-			if !slices.Contains(allowedMounts, mp.Mountpoint) {
-				return fmt.Errorf("cannot use custom mount points yet, found: %v", mp.Mountpoint)
-			}
+	return nil
+}
+
+// updateFilesystemSizes updates the size of the root filesystem customization
+// based on the minRootSize. The new min size whichever is larger between the
+// existing size and the minRootSize. If the root filesystem is not already
+// configured, a new customization is added.
+func updateFilesystemSizes(fsCustomizations []blueprint.FilesystemCustomization, minRootSize uint64) []blueprint.FilesystemCustomization {
+	updated := make([]blueprint.FilesystemCustomization, len(fsCustomizations), len(fsCustomizations)+1)
```

Comment:
(nitpick) Sorry, this maybe a silly question, but do we need to make a copy here? Or could we just modify the slice in-place (quick test below). But fine of course, if you prefer it this way (mostly curious :)
```go
diff --git a/bib/cmd/bootc-image-builder/image.go b/bib/cmd/bootc-image-builder/image.go
index 6c2c89f..1247ab8 100644
--- a/bib/cmd/bootc-image-builder/image.go
+++ b/bib/cmd/bootc-image-builder/image.go
@@ -155,21 +155,19 @@ func checkFilesystemCustomizations(fsCustomizations []blueprint.FilesystemCustom
 // existing size and the minRootSize. If the root filesystem is not already
 // configured, a new customization is added.
 func updateFilesystemSizes(fsCustomizations []blueprint.FilesystemCustomization, minRootSize uint64) []blueprint.FilesystemCustomization {
-       updated := make([]blueprint.FilesystemCustomization, len(fsCustomizations), len(fsCustomizations)+1)
        hasRoot := false
-       for idx, fsc := range fsCustomizations {
-               updated[idx] = fsc
-               if updated[idx].Mountpoint == "/" {
-                       updated[idx].MinSize = max(updated[idx].MinSize, minRootSize)
+       for idx := range fsCustomizations {
+               if fsCustomizations[idx].Mountpoint == "/" {
+                       fsCustomizations[idx].MinSize = max(fsCustomizations[idx].MinSize, minRootSize)
                        hasRoot = true
                }
        }
 
        if !hasRoot {
                // no root customization found: add it
-               updated = append(updated, blueprint.FilesystemCustomization{Mountpoint: "/", MinSize: minRootSize})
+               fsCustomizations = append(fsCustomizations, blueprint.FilesystemCustomization{Mountpoint: "/", MinSize: minRootSize})
        }
-       return updated
+       return fsCustomizations
 }
 
 // setRootfsType sets the filesystem type for the mountable entity with target

```

Reply:
I did do it like this at first, but I felt it was simpler to keep the function pure.  There's no real need to keep `customizations.Filesystem` intact currently, but I can see how modifying the user's input can get messy if for some reason we end up needing to read it after this function is called.  There's the extra issue of it not being entirely consistent.  Sometimes, what you get back will be the same object that's attached to the `customizations` and sometimes it wont (depending on whether `append()` is called or not).  I'd prefer to avoid the scenario where we modify one of the arrays after the fact and affect the other but it only happens in some scenarios and becomes a hard to trace bug.

Reply:
Thanks for the rational!

Reply:
I hope it makes sense :)